src/net/sourceforge/kolmafia/request/RelayRequest.java

```diff
@@ -3500,6 +3515,191 @@ public boolean pollForCompletion(String type) {
     return true;
   }
 
+  private void jsonError(String message) {
+    this.statusLine = "HTTP/1.1 400 Bad Request";
+    this.contentType = "application/json";
+    this.responseCode = 400;
+    this.responseText = JSON.toJSONString(Map.of("error", message));
+  }
+
+  private String underscoreName(String name) {
+    return name.replaceAll("[A-Z]", "_$0").toLowerCase();
+  }
+
+  // All existing JS type names follow this pattern
+  private boolean isJavascriptTypeNameFormat(String name) {
+    var nameUpper = name.toUpperCase();
+    var nameLower = name.toLowerCase();
+    return nameUpper.substring(0, 1).equals(name.substring(0, 1))
+        && nameLower.substring(1).equals(name.substring(1));
+  }
+
+  private Value transformApiArgument(Object thing) {
+    if (thing instanceof JSONObject obj) {
+      var objectTypeObject = obj.get("objectType");
+      Type dataType;
+      if (objectTypeObject instanceof String objectType
+          && isJavascriptTypeNameFormat(objectType)
+          && ((dataType = DataTypes.enumeratedTypes.find(objectType.toLowerCase())) != null)) {
+        var identifierStringObject = obj.get("identifierString");
+        var identifierNumberObject = obj.get("identifierNumber");
+        if (identifierNumberObject instanceof Integer identifierNumber) {
+          return dataType.makeValue(identifierNumber, true);
+        } else if (identifierStringObject instanceof String identifierString) {
+          return dataType.parseValue(identifierString, true);
+        }
+      }
+
+      return JSONValueConverter.fromJSON(obj);
+    } else return JSONValueConverter.fromJSON(thing);
```

Comment:
want a test for this -- function taking a normal value.

e.g. `numberologyPrize` should work

Reply:
this was for a function taking a non-object argument, line 3554